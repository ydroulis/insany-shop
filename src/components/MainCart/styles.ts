"use client"

import styled from 'styled-components'

export const Container = styled.main`
    display: flex;
    gap: 32px;
    font-family: Inter;
    padding: 5rem 10rem 7.188rem 10rem;

    position: relative;
`

export const Back = styled.button`
    display: flex;
    align-items: center;
    gap: 8px;

    border: none;
    background-color: transparent;
    cursor: pointer;

    font-size: 0.875rem;
    color: #617480;

    margin-top: 1.563rem;
    margin-bottom: 1.375rem;

    position: absolute;
    left: 10rem;

    &:hover {
        opacity: 0.8;
    }
`

export const Title = styled.p`
    color: #000000;
    margin: 10rem auto;
`