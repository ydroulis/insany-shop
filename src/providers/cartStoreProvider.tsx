'use client'

import { type ReactNode, createContext, useRef, useContext } from 'react'
import { useStore } from 'zustand'

import { type CartStore, createCartStore } from '../stores/cartStore'

export type CartStoreApi = ReturnType<typeof createCartStore>

export const CartStoreContext = createContext<CartStoreApi | undefined>(
    undefined,
)

export interface CartStoreProviderProps {
    children: ReactNode
}

export const CartStoreProvider = ({
    children,
}: CartStoreProviderProps) => {
    const storeRef = useRef<CartStoreApi | null>(null)
    if (storeRef.current === null) {
        storeRef.current = createCartStore()
    }

    return (
        <CartStoreContext.Provider value={storeRef.current}>
            {children}
        </CartStoreContext.Provider>
    )
}

export const useCartStore = <T,>(
    selector: (store: CartStore) => T,
): T => {
    const cartStoreContext = useContext(CartStoreContext)

    if (!cartStoreContext) {
        throw new Error(`useStore must be used within CartStoreProvider`)
    }

    return useStore(cartStoreContext, selector)
}
