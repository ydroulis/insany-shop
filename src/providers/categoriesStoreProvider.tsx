'use client'

import { type ReactNode, createContext, useRef, useContext } from 'react'
import { useStore } from 'zustand'

import { type CategoriesStore, createCategoriesStore } from '../stores/categoriesStore'

export type CategoriesStoreApi = ReturnType<typeof createCategoriesStore>

export const CategoriesStoreContext = createContext<CategoriesStoreApi | undefined>(
    undefined,
)

export interface CategoriesStoreProviderProps {
    children: ReactNode
}

export const CategoriesStoreProvider = ({
    children,
}: CategoriesStoreProviderProps) => {
    const storeRef = useRef<CategoriesStoreApi | null>(null)
    if (storeRef.current === null) {
        storeRef.current = createCategoriesStore()
    }

    return (
        <CategoriesStoreContext.Provider value={storeRef.current}>
            {children}
        </CategoriesStoreContext.Provider>
    )
}

export const useCategoriesStore = <T,>(
    selector: (store: CategoriesStore) => T,
): T => {
    const categoriesStoreContext = useContext(CategoriesStoreContext)

    if (!categoriesStoreContext) {
        throw new Error(`useStore must be used within CategoriesStoreProvider`)
    }

    return useStore(categoriesStoreContext, selector)
}
