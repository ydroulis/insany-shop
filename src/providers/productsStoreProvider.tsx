'use client'

import { type ReactNode, createContext, useRef, useContext } from 'react'
import { useStore } from 'zustand'

import { type ProductsStore, createProductsStore } from '../stores/productsStore'

export type ProductsStoreApi = ReturnType<typeof createProductsStore>

export const ProductsStoreContext = createContext<ProductsStoreApi | undefined>(
    undefined,
)

export interface ProductsStoreProviderProps {
    children: ReactNode
}

export const ProductsStoreProvider = ({
    children,
}: ProductsStoreProviderProps) => {
    const storeRef = useRef<ProductsStoreApi | null>(null)
    if (storeRef.current === null) {
        storeRef.current = createProductsStore()
    }

    return (
        <ProductsStoreContext.Provider value={storeRef.current}>
            {children}
        </ProductsStoreContext.Provider>
    )
}

export const useProductsStore = <T,>(
    selector: (store: ProductsStore) => T,
): T => {
    const productsStoreContext = useContext(ProductsStoreContext)

    if (!productsStoreContext) {
        throw new Error(`useStore must be used within ProductsStoreProvider`)
    }

    return useStore(productsStoreContext, selector)
}
