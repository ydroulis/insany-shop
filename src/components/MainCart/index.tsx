import React from 'react';

import * as S from './styles';
import CartList from '../CartList';
import SumaryOrder from '../SumaryOrder';
import { useCartStore } from '../../providers/cartStoreProvider';
import { useRouter } from 'next/navigation';
import Image from 'next/image';
import CartFeedback from '../CartFeedback';

const MainCart: React.FC = () => {
    const { cart, feedback, showFeedback } = useCartStore((state) => state);
    const router = useRouter();

    const handleBack = () => {
        showFeedback(false);
        router.back();
    }

    return (
        <S.Container
            role="main"
        >
            <S.Back data-testid="back-button" onClick={handleBack} aria-label="Voltar para a página anterior">
                <Image src="/back.png" alt="Ícone voltar" width={24} height={24} />
                Voltar
            </S.Back>
            {cart.content.length > 0 ? (
                <>
                    <CartList products={cart.content} aria-label="Lista de produtos no carrinho" />
                    <SumaryOrder subtotal={cart.total} aria-label="Resumo do pedido" />
                </>
            ) : (
                <div style={{ margin: '0 auto' }} role="status" aria-labelledby="empty-cart-title">
                    <S.Title id="empty-cart-title">Seu carrinho esta vazio</S.Title>
                </div>
            )}
            {feedback && <CartFeedback color='#DE3838' background='#fee2e2' type='error'>
                Produto removido do carrinho
            </CartFeedback>}
        </S.Container>
    );
}

export default MainCart;