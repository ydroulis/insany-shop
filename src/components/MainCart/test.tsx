import React, { FC, ImgHTMLAttributes } from 'react';
import { render, screen } from '@testing-library/react';
import MainCart from './';
import { useCartStore } from '../../providers/cartStoreProvider';
import type { CartStore } from '../../stores/cartStore';
import type { CartProduct } from '../../types/Cart';

jest.mock('../../providers/cartStoreProvider', () => ({
    useCartStore: jest.fn(),
}));

const mockBack = jest.fn();
jest.mock('next/navigation', () => ({
    useRouter: () => ({
        back: mockBack,
    }),
}));

jest.mock('next/image', () => {
    const MockedImage: FC<ImgHTMLAttributes<HTMLImageElement>> = (props) => {
        return <img {...props} alt={props.alt || 'mocked image'} />;
    };
    return {
        __esModule: true,
        default: MockedImage,
    };
});

jest.mock('../CartList', () => ({
    __esModule: true,
    default: ({ products, ...props }: { products: CartProduct[] }) => (
        <ul {...props}>
            {products.map((p) => (
                <li key={p.id} data-testid="cart-product">
                    {p.name}
                    <svg data-testid="icon-svg" />
                </li>
            ))}
        </ul>
    ),
}));

jest.mock('../SumaryOrder', () => ({
    __esModule: true,
    default: ({ subtotal, ...props }: { subtotal: number }) => (
        <section {...props}>
            <span data-testid="subtotal">{subtotal}</span>
            <svg data-testid="summary-icon" />
        </section>
    ),
}));

describe('<MainCart />', () => {
    afterEach(() => {
        jest.clearAllMocks();
    });

    const mockUseCartStore = (mockData: CartStore) => {
        (useCartStore as jest.Mock).mockImplementation(
            (selector: (state: CartStore) => unknown) => selector(mockData)
        );
    };

    it('should render cart with products', () => {
        const cartProducts: CartProduct[] = [
            { id: 1, name: 'Produto 1', price: 100, items: 1, description: '', image: '', stock: 0 },
            { id: 2, name: 'Produto 2', price: 100, items: 1, description: '', image: '', stock: 0 },
        ];

        mockUseCartStore({
            cart: { content: cartProducts, total: 200 },
            addProductToCart: jest.fn(),
            removeProductFromCart: jest.fn(),
            changeItems: jest.fn(),
            showFeedback: jest.fn(),
            feedback: false,
        });

        render(<MainCart />);

        expect(screen.getByRole('main')).toBeInTheDocument();
        expect(screen.getAllByTestId('cart-product')).toHaveLength(2);
        expect(screen.getAllByTestId('icon-svg')).toHaveLength(2);
        expect(screen.getByTestId('subtotal')).toHaveTextContent('200');
        expect(screen.getByTestId('summary-icon')).toBeInTheDocument();
    });

    it('should render empty cart message', () => {
        mockUseCartStore({
            cart: { content: [], total: 0 },
            addProductToCart: jest.fn(),
            removeProductFromCart: jest.fn(),
            changeItems: jest.fn(),
            showFeedback: jest.fn(),
            feedback: false,
        });

        render(<MainCart />);

        const status = screen.getByRole('status');
        expect(status).toBeInTheDocument();
        expect(screen.getByText(/seu carrinho esta vazio/i)).toBeInTheDocument();
        expect(status).toHaveAttribute('aria-labelledby', 'empty-cart-title');
    });

    it('should match snapshot', () => {
        const cartProducts: CartProduct[] = [
            { id: 1, name: 'Produto 1', price: 100, items: 1, description: '', image: '', stock: 0 },
            { id: 2, name: 'Produto 2', price: 100, items: 1, description: '', image: '', stock: 0 },
        ];

        mockUseCartStore({
            cart: { content: cartProducts, total: 200 },
            addProductToCart: jest.fn(),
            removeProductFromCart: jest.fn(),
            changeItems: jest.fn(),
            showFeedback: jest.fn(),
            feedback: false,
        });

        const { container } = render(<MainCart />);
        expect(container).toMatchSnapshot();
    });
});
