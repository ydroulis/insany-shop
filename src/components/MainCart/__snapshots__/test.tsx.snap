// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<MainCart /> should match snapshot 1`] = `
<div>
  <main
    class="sc-bRKDuR igzaSt"
    role="main"
  >
    <button
      aria-label="Voltar para a página anterior"
      class="sc-hvigdm jSpvAU"
      data-testid="back-button"
    >
      <img
        alt="Ícone voltar"
        height="24"
        src="/back.png"
        width="24"
      />
      Voltar
    </button>
    <ul
      aria-label="Lista de produtos no carrinho"
    >
      <li
        data-testid="cart-product"
      >
        Produto 1
        <svg
          data-testid="icon-svg"
        />
      </li>
      <li
        data-testid="cart-product"
      >
        Produto 2
        <svg
          data-testid="icon-svg"
        />
      </li>
    </ul>
    <section
      aria-label="Resumo do pedido"
    >
      <span
        data-testid="subtotal"
      >
        200
      </span>
      <svg
        data-testid="summary-icon"
      />
    </section>
  </main>
</div>
`;
